linkerd/meshtls/src/client.rs

```diff
@@ -0,0 +1,200 @@
+        if let Self::Rustls(connect) = self {
+            return <rustls::Connect as Service<I>>::poll_ready(connect, cx);
+        }
+
+        unreachable!()
+    }
+
+    #[inline]
+    fn call(&mut self, io: I) -> Self::Future {
+        #[cfg(feature = "rustls")]
+        if let Self::Rustls(connect) = self {
+            return ConnectFuture::Rustls(connect.call(io));
+        }
+
+        unreachable!()
+    }
+}
+
+// === impl ConnectFuture ===
+
+impl<I> Future for ConnectFuture<I>
+where
+    I: io::AsyncRead + io::AsyncWrite + Unpin,
+{
+    type Output = io::Result<ClientIo<I>>;
+
+    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        #[cfg(feature = "rustls")]
+        if let ConnectFutureProj::Rustls(f) = this {
+            let res = futures::ready!(f.poll(cx));
+            return Poll::Ready(res.map(ClientIo::Rustls));
+        }
+
+        unreachable!()
+    }
+}
+
+// === impl ClientIo ===
+
+impl<I: io::AsyncRead + io::AsyncWrite + Unpin> io::AsyncRead for ClientIo<I> {
+    #[inline]
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut io::ReadBuf<'_>,
+    ) -> io::Poll<()> {
+        let this = self.project();
+
+        #[cfg(feature = "rustls")]
+        if let ClientIoProj::Rustls(io) = this {
+            return io.poll_read(cx, buf);
+        }
+
+        unreachable!()
+    }
+}
+
+impl<I: io::AsyncRead + io::AsyncWrite + Unpin> io::AsyncWrite for ClientIo<I> {
+    #[inline]
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> io::Poll<()> {
+        let this = self.project();
+
+        #[cfg(feature = "rustls")]
+        if let ClientIoProj::Rustls(io) = this {
+            return io.poll_flush(cx);
+        }
+
+        unreachable!()
+    }
+
+    #[inline]
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> io::Poll<()> {
+        let this = self.project();
+
+        #[cfg(feature = "rustls")]
+        if let ClientIoProj::Rustls(io) = this {
+            return io.poll_shutdown(cx);
+        }
+
+        unreachable!()
+    }
+
+    #[inline]
+    fn poll_write(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> io::Poll<usize> {
+        let this = self.project();
+
+        #[cfg(feature = "rustls")]
+        if let ClientIoProj::Rustls(io) = this {
+            return io.poll_write(cx, buf);
+        }
+
+        unreachable!()
+    }
+
+    #[inline]
+    fn poll_write_vectored(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[io::IoSlice<'_>],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        let this = self.project();
+
+        #[cfg(feature = "rustls")]
+        if let ClientIoProj::Rustls(io) = this {
+            return io.poll_write_vectored(cx, bufs);
+        }
+
+        unreachable!()
+    }
+
+    #[inline]
+    fn is_write_vectored(&self) -> bool {
+        unimplemented!()
```

Comment:
the rustls impl should probably implement this?
```suggestion
        #[cfg(feature = "rustls")]
        if let ClientIo::Rustls(ref io) = self {
            return io.is_write_vectored();
        }
        
        unreachable!()
```